chainermn/communicators/mpi_communicator_base.py

```diff
@@ -0,0 +1,375 @@
+        chainer.Variable objects. Please be sure.
+
+        Args:
+            data: data to be sent (tuple, list or raw numpy/cupy array)
+            dest (int): Target process specifier.
+            tag (int): Message ID (MPI feature).
+
+        """
+        chainer.utils.experimental(
+            'chainermn.communicators.CommunicatorBase.send')
+
+        msgtype = _MessageType(data)
+        """We use ssend() instead of send() to pass unittests.
+        If we don't use it, an error occurs in
+        test_point_to_point_communication.py
+        when using MVAPICH2-2.2 and GPUs.
+        """
+        self.mpi_comm.ssend(msgtype, dest=dest, tag=tag)
+
+        # Type check.
+        if not msgtype.is_tuple:
+            data = [data]
+
+        for x in data:
+            if x.dtype != numpy.float32:
+                raise ValueError('send only support dtype == numpy.float32')
+
+        for array in data:
+            if chainer.cuda.get_array_module(array) is not numpy:
+                chainer.cuda.Stream.null.synchronize()
+
+            buf = _memory_utility.array_to_buffer_object(array)
+            """We use Ssend() for the same reason as using ssend()."""
+            self.mpi_comm.Ssend(buf, dest=dest, tag=tag)
+
+    def recv(self, source, tag):
+        """A primitive of inter-process receiver.
+
+        This method tries to receive numpy-array from target process.
+        The target process is expected to invoke ``send()``.
+        This method relies on mpi4py fast communication optimized for
+        numpy arrays, which discards any information attached to
+        chainer.Variable objects. Please be sure.
+
+        Args:
+            source (int): Target process specifier.
+            tag (int): Message ID (MPI feature).
+
+        """
+
+        chainer.utils.experimental(
+            'chainermn.communicators.CommunicatorBase.recv')
+
+        msgtype = self.mpi_comm.recv(source=source, tag=tag)
+
+        if msgtype.is_tuple:
+            msg = []
+            for shape in msgtype.shapes:
+                buf = numpy.empty(numpy.prod(shape), dtype=numpy.float32)
+                self.mpi_comm.Recv(buf, source=source, tag=tag)
+                msg.append(buf.reshape(shape))
+            return tuple(msg)
+
+        else:
+            assert len(msgtype.shapes) == 1
+            shape = msgtype.shapes[0]
+            buf = numpy.empty(numpy.prod(shape), dtype=numpy.float32)
+            self.mpi_comm.Recv(buf, source=source, tag=tag)
+            return buf.reshape(shape)
+
+    def bcast(self, x, root=0):
+        """A primitive of inter-process broadcast communication.
+
+        This method tries to invoke broadcast communication within the
+        communicator. All processes in the communicator are expected to
+        invoke ``broadcast()``. This method relies on mpi4py fast communication
+        optimized for numpy arrays, as well as ``send()`` and ``recv()``.
```

Comment:
I didn't quite understand the meaning of `as well as send() and recv()`. 
Does it mean...

(1) the method relies on mpi4py's fast communication, mpi4py's send(), and mpi4py's recv()
(Does it really rely on mpi4py's send() and recv() ?)

OR,

(2) This method relies on mpi4py fast communication ...,  as send() and recv() does?

Reply:
These methods are all copied from MPI part of  `_base.py`.

Reply:
Got it!